### ISSUE TITLE
MetadataViews.SerialNumber

### ISSUE THREAD

**Issue by poiati** [onflow/flow-nft#58]
### Issue To Be Solved

Now that we have `MetadataViews.Display`, the next piece of data which is present in a significant portion of the NFT's out there is the so-called `Serial Number`. Unlike name and description, this information is not appliable to all tokens. That said, it's still something people care about. 

### Solution 1: MetadataViews.SerialNumber

Introduce the `MetadataViews.SerialNumber` view:

```
// ... MetadataViews Contract ...

pub struct SerialNumber {
  pub let serialNumber: UInt64
  pub let totalMinted: UInt64? // Not sure about this
}
```

PS: I'm still unsure if we should include the total number of minted NFT's. E.g: #1/200 (1 out of 200 copies). Mainly because this information is more likely to change.

### Solution 2: Serial number as part of the name / description

_This is not a solution, just an alternative._ 

If we don't want the Serial Number as first-class information within `MetadataViews` we can include it as part of the name/description in the `Display` view. It would be complex and error-prone if the client wanted to extract this information.

**Comment by joshuahannan** [comment]
Do we think that the serial number can always be enforced to be a UInt64?

I like having a specific view for serial number, but I don't think it should be included in Display

**Comment by bluesign** [comment]
I agree with @joshuahannan, I think `Display` should be as thin as possible.

**Comment by poiati** [comment]
Yes, what I'm suggesting is a view apart from `Display`.

> Do we think that the serial number can always be enforced to be a UInt64?

I would say yes, @joshuahannan . What are your concerns? Is it about the size? Have you seen any other types being used? I see benefits from a dev perspective in having a homogenous representation. 

> @poiati did I understand your intention here or is this something else?

Yes, you nailed it @bjartek.

**Comment by bjartek** [comment]
I think then that Solution 1 is the way to go, but I am not sure on the naming of the struct or the fields. I think `Editions` with the fields `edition` and `totalEdition`/`maxEdition` is more clear here. Just my 2 cents.

**Comment by joshuahannan** [comment]
`UInt64` is totally fine with me. I just was thinking that there might be a chance that other projects would want something else, but I don't think it really matters. 

As for the name of the struct, I think `SerialNumber` is probably fine, but the second field should probably be something like `totalInEdition.` Though that is the first mention of Edition without any context, so maybe it isn't great.

**Comment by SeanRobb** [comment]
Can we consider editions to have an `id`, `name`, and `description`?  This way I could see if NFTs are similar groupings of each other?

**Comment by bjartek** [comment]
@SeanRobb In my eyes if you want that you fetch both the Display view and the SerialNumber view from the same resource and then you have that. We should not duplicate information in views that are already present in other core views unless strictly needed.

**Comment by SeanRobb** [comment]
@bjartek Completely agree. After talking more with @poiati I initially thought this could be coupled in a view but this is incorrect. 

I am really looking for a way for complex NFT concepts to be grouped together to offer or buy sets of NFT's that can be fulfilled by one NFT.

For example, if I want any NFT from the `Ballerz` set from [Gaia](https://flowscan.org/contract/A.8b148183c28ff88f.Gaia/overview) or any [NBA Top Shot](https://flowscan.org/contract/A.0b2a3299cc857e29.TopShot/overview) from a certain `Set` or `Play` it would be nice to allow for a Metadata View so that NFT creators can group these logically allowing for `Sales` and `Offers` to potentially be grouped as well.   I will write up a separate issue for this functionality and link it back to this comment.

**Comment by bjartek** [comment]
Totally agree, create a seperate issue for them?

**Comment by joshuahannan** [comment]
Anyone have any more thoughts on this? @bjartek Do you want to create a PR for your view also?

**Comment by bjartek** [comment]
I can create one for sure, if nobody else has any input on it.

**Comment by poiati** [comment]
I personally don't like attaching the name `Edition` here, especially if the name of the view is kept as `SerialNumber`. I see an edition as a higher-level abstraction, I also find it harder to generalize. For example, If we want to semantically represent an Edition we probably also want to have a name there. 

I would keep this as simple as possible like my original proposal, or change to something with the same structure/semantics but different naming.

E.g:

```
pub struct Serial {
  pub let serialNumber: UInt64
  pub let cap: UInt64?
}
```

**Comment by joshuahannan** [comment]
I agree, that keeping it simple is definitely good. Naming is hard though. I don't have much insight on what a good name would be for the view and fields though. @poiati Yours seems fine, but I don't feel strongly about that

**Comment by poiati** [comment]
@joshuahannan, how does this PR fixes the issue?

**Comment by poiati** [comment]
I disagree #75 solves this issue. If it does, it does in a very confusing way. Most of the third-party project we onboarded in Dapper is generative NFT collections. They are not an `Edition`, that's why I'm strongly suggesting to not use that name in this specific context.

```
Edition information is very useful in wallets and marketplaces. As more and more projects on Flow mint multiple items with the same name, description and media file, it is beneficial to put this into standard.
```
This just isn't true. The name, description and media can and usually vary for those projects (Ballerz, Flunks...). Just take Opensea for example, it's called Serial Number, people with the context in crypto are used to it. I don't understand why we would deviate from that.

The `Edition` view is useful, but not here IMO.

**Comment by poiati** [comment]
@joshuahannan, we can technically do that, but think about how confusing to the developer it is. The problem here is about semantics, not structure. If I go through docs, I will find that the `Edition` reads as:
 
 ```
Edition information is very useful in wallets and marketplaces. As more and more projects on Flow mint multiple items with the same name, description and media file, it is beneficial to put this into standard.
```

This makes sense since, in the real world, an edition semantically represents a collection of the same thing. Generative projects are clearly not the same, they have a different set of attributes, and rarity.

I wouldn't be associating my generative project with it all. Don't you agree? Editions make a lot of sense in our Sports projects. It doesn't for many other projects though. I'm not saying it's unusable, I'm saying I don't think this is the best option.

Maybe the problem is that those generative projects shouldn't be using the term "Serial Number" then, the fact is, most of them do. And this is a general concept accepted whatever blockchain you are using.

It's all about making it clear and easy for the developer, using the standards they are used to.

**Comment by poiati** [comment]
@bjartek SerialNumber is an abstraction (a view), its concrete implementation might be anything the developer chooses, including the NFT ID.

**Comment by poiati** [comment]
Everything that I'm saying is based on empirical observation in projects we onboarded in Dapper. I can say the same thing for Ethereum/Opensea.

I can give you examples:

https://flow-view-source.com/mainnet/account/0x807c3d470888cc48/contract/Flunks
https://flow-view-source.com/mainnet/account/0x8b148183c28ff88f/contract/Gaia

**Comment by bjartek** [comment]
I am not confused anymore. What you are suggestion is not what I thought you where suggesting. So i misunderstood. I have no problem with a SerialNumber view to cover this scenario.

**Comment by joshuahannan** [comment]
I still think that this and the edition views should be combined. Would it make you feel better if we just used your naming for it instead of what Hsuan proposed?
```
pub struct Serial {
  pub let serialNumber: UInt64
  pub let cap: UInt64?
}
```

**Comment by bjartek** [comment]
From my perspective editions/maxEditions and serial number is two different things. 

Editions are for things that are identical except the editions. Where as serialNumber are all different. It is just another identifier for an nft.

**Comment by poiati** [comment]
I like the `Serial` interface. It's very minimal, flexible, and solves the problem. For example, an Edition can be thought of as a superset of the `Serial` view.

I'm not sure about the `Edition` interface of #75. I feel like we would like to have more information to define it. For example, how do I know if two distinct NFTs belong to the same edition? There is no way to do that through this view, which is weird IMO, should we also have an identifier for the Edition?

**Comment by poiati** [comment]
I don't want to be a blocker here. Everything I am saying is based on contracts I read, and every single one (TopShot, AllDay, Genies...) has its own way of defining an Edition.

TopShot and AllDay use the term `Play` while Genies use `Edition`, Gaia names it `Template`. All of them have one thing in common, an identifier.

I wonder based on what evidence we got to the `Edition` definition we have on #75. Personally, I would say it's too early to call for it. We need more information.

The `Serial` interface partially solves the problem by enabling Dapps to sequentially identify their NFT's. We still can't tell if there are more identical NFTs though. Which is what @bjartek was saying about their difference.

I agree with @joshuahannan that it isn't a good idea to introduce both interfaces as it is. IMO though, the problem is with the Edition interface.

**Comment by joshuahannan** [comment]
@poiati Can you leave these comments on that PR so the author and others can comment on it also? I think it is an interesting conversation, but we need more input on it.

**Comment by bjartek** [comment]
I looked at topshot moment and on cryptoslam they have both sn# and serial number and edition size.

Sn# looks to be id. Serial number looks to me to by what in the art nft scene is edition and editionSize is then maxEdition.

My opinion:
- edition or serialNumber i do not care
- i do not like the cap name. It means capability in my book.
- editionSize is better, but totalSerial works to.

**Comment by poiati** [comment]
I agree cap is context-sensitive and might look ambiguous. I'm fine as long we keep the term `edition` out of it.

**Comment by poiati** [comment]
We could also use `max` or `limit`.

**Comment by boczeratul** [comment]
I'm ok with renaming it to `Serial` so it can serve both cases: a set of entirely identical items and a set of similar items (as in PFP projects like Flovatar)

I don't think we should include Serial identifier though. It can vary a lot from project to project. Some may use `UInt`s, some may use `UUID`s, while some may use `String`s to identify their group.

**Comment by bluesign** [comment]
What if someone make a PFP but each one is limited to 5 only instead of 1. 

so I can have an item 42/1000 and 2/5 

This looks more like a Book. It has editions with different content, but serial within edition.

**Comment by joshuahannan** [comment]
What does everyone think about @bluesign's proposal? Do we think that it can make serials and editions work together or do we think they should be separate?

**Comment by bjartek** [comment]
It works for me.

**Comment by tyllenb** [comment]
@poiati how does @bluesign proposal sound for you? I think this is a nice implementation of it was well

**Comment by poiati** [comment]
I still don't think we should touch this interface. IMO the changes are confusing. `name` does not fit here.

> As it has been used different with terminology across projects and Items can have many Serials/Editions or other grouping.

How would that be possible? `resolveView` accepts a type as an argument to retrieve the metadata. AFAIK we can't have two implementations of the same view type.

> that way if Versus using it they can name it Edition, another project can name it Serial.

What are other possible names? If we are limiting to those two I don't think adding this information makes sense.

```
Also if project has 2 Serial, it can also return 2 Serial View with different names, in my PFP example can be:

Serial(name=Edition, serialNumber=42, maxNumber=1000)
and
Serial(name=Serial, serialNumber=2, maxNumber=5)
```

I don't think this is possible for the same reasons stated above.

**Comment by bjartek** [comment]
> How would that be possible? resolveView accepts a type as an argument to retrieve the metadata. AFAIK we can't have two implementations of the same view type.
You can expose a list of a type to signal that you have more of them. Or a dictionary. 

> What are other possible names? If we are limiting to those two I don't think adding this information makes sense.
The possible names are infiite here, if you want Serials you should ask for the view and check the the name is Serial?

> I don't think this is possible for the same reasons stated above.
I have tried to show that it is possible. One project can expose Edition and solutions that want editions can filter on that and others can filter on Serial. 

I think this is very elegant solution. For instance topshot could here expose a list of Serial and all one of the Set/Season/Serial(even though it is called edition in the UI).

**Comment by boczeratul** [comment]
The proposal from @bluesign looks good to me.
If we're ok with this, I can modify the PR #75 to this proposal.

**Comment by tyllenb** [comment]
@boczeratul good to modify PR #75

**Comment by poiati** [comment]
> You can expose a list of a type to signal that you have more of them. Or a dictionary.

I see, thanks for bringing this. IMO this adds complexity though, not sure if I like it. For Dapper we just need so show the serial number, to do that now I need to do loosy string comparations to a "name".

> The possible names are infiite here, if you want Serials you should ask for the view and check the the name is Serial?

I see this answer more like a cons than a pro. Abstraction should enforce semantics, having it as loosely defined "the sky is the limit" is usually not a good idea. Can you bring another concrete example aside from Serial and Edition?

If what we are pushing here is the ultimate generalization (I'm not) we can do a `Meta` view which is just a key-value pair. We can represent Editions, Serials or whatever we want:

```
key: "Serial"
value: "#1/#10000"
```

**Comment by bjartek** [comment]
> Can you bring another concrete example aside from Serial and Edition?

Set, Season

> Abstraction should enforce semantics, having it as loosely defined "the sky is the limit" is usually not a good idea. 

I totally agree, and in that though it Serials, Editions and OpenEditions are different things then just create a _simple_ struct for each of them. 

 - Serial(number:Int)
 - Edition(number:Int, max:Int) //or even add a name to say what the edition is about, but in my eyes it is everything but edition is the same.
 - OpenEdition(number:Int)

I prefer and like simple things that compose and you can combine them. But earlier in this thread when I suggested having both Editions and Serials that was not accepted.

**Comment by poiati** [comment]
> Set, Season

Make sense. Thank you.

> I totally agree, and in that though it Serials, Editions and OpenEditions are different things then just create a _simple_ struct for each of them.
>
> * Serial(number:Int)
> * Edition(number:Int, max:Int) //or even add a name to say what the edition is about, but in my eyes it is everything but edition is the same.
> * OpenEdition(number:Int)
> 
> I prefer and like simple things that compose and you can combine them. But earlier in this thread when I suggested having both Editions and Serials that was not accepted.

I'm not sure why that wasn't accepted. All my arguments were against missing those two concepts together. Even if they need to have the same structure, I think what you are suggesting is far better than the `Serial` interface lastly defined here.

**Comment by poiati** [comment]
If it's a simple PFP or a single edition it just implements `Serial`. If it has Serial in the context of editions it implements both.

**Comment by bluesign** [comment]
Problem is what we call `Serial` here can be anything depending on the domain. 

If someone is selling NFT of Books, Music, Art, Trading Cards, Avatars they all have different meaning for `Serial` / `Edition` on context. 

Basic properties are: 

- Sequential / Non Sequential ( Category / Label ) 
- Limited / Unlimited 
- Order has importance / Order does not have importance 
- etc

Also Item's can have more than one `Serial`, like TopShot NFT having: Season, Set, Play, Serial 

Although it seems to be tempting to use Typed views, it will end up people using them for different meanings I am afraid. It is better to let people define their properties then forcing them to fit their domain in to some restricted space.

**Comment by poiati** [comment]
> Problem is what we call `Serial` here can be anything depending on the domain.

Definition of Serial in the context of this issue:
Serial is a sequence of numbers where if I get two NFTs and both have the same serial they MUST BE the same NFT. In other words, the serial can, but don't need to be, the token ID.

> Basic properties are:
> 
> * Sequential / Non Sequential ( Category / Label )

AFAIK there isn't such a thing as non-sequential serial, thus the name. I might be wrong though.

> * Limited / Unlimited

Thus the `cap` or `max` attribute.

> * Order has importance / Order does not have importance

Why does that matter?
 
> Although it seems to be tempting to use Typed views, it will end up people using them for different meanings I am afraid. It is better to let people define their properties then forcing them to fit their domain in to some restricted space.

I can easily document the `Serial` view here, without much effort I can set clear where and when it should be used. I have a hard time with the last proposal, just because it can basically fit anywhere.

IMO the danger is in not restricting. The view doesn't provide much value if we cannot determine semantic behaviors based on them. We are not preventing them to create their own view types, they can totally do that.

**Comment by bluesign** [comment]
> AFAIK there isn't such a thing as non-sequential serial, thus the name. I might be wrong though.

I mean on database you have sequential number, but actually meaningful part is the `label`.

- For example TopShot `Play`/`Set` is an `Edition` by definition. But number or order has no importance.  You would not want to see `Set: 45/287`

My point is, we have 2 options: 

- We can cover most cases with a single view ( repeated if necessary ) 
- We can try to define all cases separate views.

For me first one is the clear winner, most projects will fit with this combinations. 

If we make separate Types, you need to check which type NFT implemented, take action depending on the Type. 

One will use Serial Type, another will use Edition type. ( Now terminology is not crystal clear here, Edition = Serial is interchanged unfortunately )

**Comment by austinkline** [comment]
Just got caught up! I like @bluesign's proposal. My only suggestion is to rename `maxNumber` to `totalSupply` given how well-known the term is. A maximum might not be right in general either depending on how we're displaying. Currently, a maximum implies that that's how many exist. 

For instance, if I have a PFP which is serial 100/1000 where max is 1000, I am assuming there are 1000 in circulation. But What if it doesn't mint all 1000 and there are only 250? That distinction is very important in my opinion. There should be a way to express that, and a max will lead developers to setting it to the sell-out maximum not the total in circulation (or total that got minted)

**Comment by poiati** [comment]
I disagree. The suggestion makes my use case, which made me create this issue in the first place, much harder and error-prone. All I need to do is: "Get the serial number of an NFT". Where before I had to just resolve the type, now I need to get multiple views, do string comp and have faith they have spelled the `name` as I expect, whereas before I would get compile-time errors.

I really feel like we are mixing a lot of concepts that shouldn't be mixed here. You are suggesting to include `name` and `serialDisplay`. So we can do something like:

```
Set: Super Set
Edition: #10
Serial: #34/100
```

Right? If yes, why we just don't introduce a Meta{Key, Value} view. We would be able to achieve the same result without the confusion.

**Comment by austinkline** [comment]
> If yes, why we just don't introduce a Meta{Key, Value} view

@poiati Do you mean a general-purpose Metadata view? We've discussed a little bit of that in #73

**Comment by bluesign** [comment]
> All I need to do is: "Get the serial number of an NFT". Where before I had to just resolve the type, now I need to get multiple views, do string comp and have faith they have spelled the name as I expect, whereas before I would get compile-time errors.

Yeah but what if you are just trying to get serial number of the NFT, but it is using `Edition` instead or `OpenEdition` ?

**Comment by poiati** [comment]
I don't care, that's the power of assertive abstraction. They are providing the serial by implementing the view, I don't care what it is. If they don't implement it, I assume they don't have a serial.

**Comment by poiati** [comment]
For my use case (Dapper) I don't care about editions, set, plays...

**Comment by bluesign** [comment]
Btw what is your use case for `Serial` ? So you will have requirement as `Serial` should be unique ?

What if there are `editions` (you don't care but can be) ? So you will have same serial duplicate ?

**Comment by poiati** [comment]
> Btw what is your use case for Serial ?

People want to know their serial number in Dapper. Half of the projects that Dapper supports are PFP. Even when they are not, like NBATS, people care much more about serial number and rarity than they do about Set or Edition. 

> What if there are editions (you don't care but can be) ? So you will have same serial duplicate ?

> Edition:1 Serial: 77
> Edition:10 Serial: 77

In that case, the serial will be the same in the context of the edition, which is fine. Your proposal also does not solve this problem since there is no relationship or hierarchy definition.

> those are just Serial: 77 for you ?

Yes.

**Comment by austinkline** [comment]
I'm not onboard with edition being included here, it isn't complete and it forces a single-tier hierarchy that isn't true even now. Serial view having a single edition field means I can't describe the nuance of TopShot accurately. I would now need to ask for more views to contextualize and those other views will look very similar to this one. 

To me, edition is either a separate view, or this view is somewhat generic to allow the expression of `X/Y` with a `label` where a resource can vend many of them with the needed context. 

> Even when they are not, like NBATS, people care much more about serial number and rarity than they do about Set or Edition.

Anecdotally, I'm not in agreement with this. Lenders on flowty care very much about the set and series, especially Series 1 over all others. It's a very important component of the NFT in that it distinguishes them as the genesis moments. A single edition would preclude us from expressing that wouldn't it? And if edition is the play (or the set), what makes that field the "right" one for edition to be set to versus the others?

**Comment by poiati** [comment]
> I'm not onboard with edition being included here, it isn't complete and it forces a single-tier hierarchy that isn't true even now. Serial view having a single edition field means I can't describe the nuance of TopShot accurately. I would now need to ask for more views to contextualize and those other views will look very similar to this one.

> To me, edition is either a separate view, or this view is somewhat generic to allow the expression of X/Y with a label where a resource can vend many of them with the needed context.

I'm not 100% sure what you mean but I totally onboard that `edition` is not part of this issue.

> Anecdotally, I'm not in agreement with this. Lenders on flowty care very much about the set and series, especially Series 1 over all others. It's a very important component of the NFT in that it distinguishes them as the genesis moments. A single edition would preclude us from expressing that wouldn't it? And if edition is the play (or the set), what makes that field the "right" one for edition to be set to versus the others?

The edition/set/series problem is a much more complex than what I'm proposing here. I don't think we should try to solve all problems in a single view. When I say people care more about serial I do for one reason. They can usually infer the information about the set/play from the token media/title/description (Display view). This is mostly not true for the serial.

**Comment by austinkline** [comment]
It sounds like your take on this is that Serial is it's own view and the others get figured out later. Maybe in the form of an `Edition` view where this weirdness is explicitly addressed.

> They can usually infer the information about the set/play from the token media/title/description (Display view). This is mostly not true for the serial.

Good point! Especially since this view is mostly for displaying purposes to end users versus for dapp use in transactions.

**Comment by poiati** [comment]
> It sounds like your take on this is that Serial is it's own view and the others get figured out later. Maybe in the form of an Edition view where this weirdness is explicitly addressed.

This is exactly what I would do.

**Comment by bluesign** [comment]
> When I say people care more about serial I do for one reason. They can usually infer the information about the set/play from the token media/title/description (Display view). This is mostly not true for the serial.

https://github.com/dapperlabs/nba-smart-contracts/blob/95fe72b7e94f43c9eff28412ce3642b69dcd8cd5/contracts/TopShot.cdc#L596

I really don't understand and actually gave up, you said `name` parameter is loose string comparison, and now here defending people inferring information from media/title/description

In my understanding MetadataViews are not for `People` to infer anything, it is for machines ( Wallets, Dapps, Exchanges etc ) to talk with each other.

**Comment by bjartek** [comment]
Lively discussion here. What about this?
`Serial(name: String, serial:Int, total:Int?)`

That should cover pretty much any usecase we have discussed _if_ we do not want specific views for the different types I think?

 I can see why a plain Edition without describing what it is an edition on can be confusing. From Versus/art is was always the reprint. It makes sense there because the context is Art.

**Comment by poiati** [comment]
> and now here defending people inferring information from media/title/description

@bluesign, with all due respect, can you quote me where I'm saying this?

I will try to be as short and as concise as possible because I feel like there is a lot of confusion in this conversation.

> In my understanding MetadataViews are not for People to infer anything, it is for machines ( Wallets, Dapps, Exchanges etc ) to talk with each other.

Let's shift the focus to the `Display` view for a moment. It's a general-purpose view so an NFT can be identified. This is not for machines to identify it, it's built for people, ultimately everything we are doing is writing software for them, to add value to the product. Yes, a machine consumes it, but who is taking advantage of it, in the end, are the users. There is no point in having a `Display` view if no one is seeing it.

The view does not enforce what should be the `title` and `description`, and IMO that's correct. It shouldn't. Some apps might include the Serial number there, others might not.

I 100% agree with you that we shouldn't be parsing values from the `Display` view in order to get anything, that's not what I'm suggesting. All I'm saying is that our end users are usually happy when they can find the serial there, if they can't find it, they ask our support team how they can see it. Our answer is: you can't. Every single app Dapper supports has a "Serial" labeled information, with no exceptions. Less than half has the concept of `Edition`, and `Play` is something we only see in Dapper Sports projects.

The point of the view I'm suggesting in this issue is to answer the following question:

`What is the Serial Number of a given NFT?`

That's all I need to know, that's simple and easy to document and implement. A lot of projects don't need the complexity of NBATS, this is a factual observation.

Now, going back to your thoughts:

> In my understanding MetadataViews are not for People to infer anything, it is for machines

Exactly, Isn't it better for a machine to be assertive? With what I'm proposing there is no ambiguity, which is an issue for computers.

- Does the NFT resolve the view `Serial`? If yes, we are assured to have it. If not, totally fine as well, this is an opt-in view. You don't need to implement it at all.

With your proposal, as I understand, the job is harder.

- First I need to query for a collection of views instead of just one
- Then I need to iterate and do a string comparison. Whoever is implementing this needs to make sure they are using this string, documentation becomes harder for the writer and the reader.
- Even if we are using dicts, we need to make sure we are doing the lookup with the right case, on top of that, we are only considering the English lang, not a big issue at all, but things start to sum up and the solution becomes more error-prone.

You can argue the cons are paid off given the pros of the flexibility. That's where I personally get confused. What pros? Given the developer can put any arbitrary `name` there. How computers would take proper advantage of that? The only use case I can think of is displaying, and, if that's the case. Why just not use a more generic view abstraction where we represent labels and values instead?

Last but not least, I want to enforce this is my personal thoughts and I won't be upset if we chose a solution that doesn't match my expectations. What I'm trying to do is add food for thought because what we are setting here is a big deal and will be used for many years to come.

I hope I was able to be clear. Cheers!

**Comment by poiati** [comment]
> Serial(name: String, serial:Int, total:Int?)

> That should cover pretty much any usecase we have discussed if we do not want specific views for the different types I think?

How don't want to use specific views?

**Comment by bjartek** [comment]
I do not know at this point @poiati.

**Comment by joshuahannan** [comment]
Trying to get this moving along again. After re-reading everything again, I think that I agree with @poiati that we should just do a simple SerialNumber view that only includes the serial number, then we could push some of this complexity that we have been discussing to the edition view. I think @bluesign's proposal probably makes more sense as a proposal for Editions because those do tend to have more complexity.

If nobody has any issues with that proposal, I'll make a PR with this simple `SerialNumber` view, and once that is finished, I can modify Hsuan's PR to include this proposal with something that I think will fit and we could move on from there. Hopefully that will satisfy the different use cases?

**Comment by poiati** [comment]
Agreed. I like the idea of having niche views. Thank you.

**Comment by joshuahannan** [comment]
I made a PR for this view if anyone wants to review it: https://github.com/onflow/flow-nft/pull/84
